db/note-sql_test.go

```diff
@@ -0,0 +1,293 @@
+package db
+
+import (
+	"database/sql"
+	"fmt"
+	"math/rand"
+	"testing"
+	"time"
+
+	"github.com/DATA-DOG/go-sqlmock"
+	"github.com/chadweimer/gomp/models"
+	gomock "github.com/golang/mock/gomock"
+)
+
+func Test_Note_Create(t *testing.T) {
+	type testArgs struct {
+		recipeId      int64
+		text          string
+		dbError       error
+		expectedError error
+	}
+
+	// Arrange
+	tests := []testArgs{
+		{1, "My note", nil, nil},
+		{0, "", sql.ErrNoRows, ErrNotFound},
+		{0, "", sql.ErrConnDone, sql.ErrConnDone},
+	}
+	for i, test := range tests {
+		t.Run(fmt.Sprint(i), func(t *testing.T) {
+			// Arrange
+			ctrl := gomock.NewController(t)
+			defer ctrl.Finish()
+
+			db, dbmock := getMockDb(t)
+			defer db.Close()
+			sut := sqlNoteDriver{db}
+
+			note := &models.Note{RecipeId: &test.recipeId, Text: test.text}
+			expectedId := rand.Int63()
+
+			dbmock.ExpectBegin()
+			query := dbmock.ExpectQuery("INSERT INTO recipe_note \\(recipe_id, note\\) VALUES \\(\\$1, \\$2\\) RETURNING id").WithArgs(note.RecipeId, note.Text)
+			if test.dbError == nil {
+				query.WillReturnRows(sqlmock.NewRows([]string{"id"}).AddRow(expectedId))
+				dbmock.ExpectCommit()
+			} else {
+				query.WillReturnError(test.dbError)
+				dbmock.ExpectRollback()
+			}
+
+			// Act
+			err := sut.Create(note)
+
+			// Assert
+			if err != test.expectedError {
+				t.Errorf("expected error: %v, received error: %v", ErrNotFound, err)
+			}
+			if err := dbmock.ExpectationsWereMet(); err != nil {
+				t.Errorf("there were unfulfilled expectations: %s", err)
+			}
+			if err == nil && *note.Id != expectedId {
```

Comment:
Should use test.expectedError since err is reassigned above

```suggestion
			if test.expectedError == nil && *note.Id != expectedId {
```